File: src/graph/lowering/position_embedding.cpp
/* Copyright (c) 2025-2026 BAAI. SPDX-License-Identifier: Apache-2.0 */
#include "graph/lowering/lowering.hpp"
#include "graph/lowering/helpers.hpp"
#include <cmath>
namespace flagdnn::native {
LoweredOperation lower_rope(const OperationSpec& operation) {
  const bool backward = operation.custom_operation_name == "rope_backward";
  require_port_count(operation, 2, 1);
  const auto& x =
      require_port(operation.inputs, backward ? "dy" : "input", "input");
  const auto& freqs = require_port(operation.inputs, "freqs", "input");
  const auto& y =
      require_port(operation.outputs, backward ? "dx" : "output", "output");
  for (const auto* tensor : {&x, &freqs, &y})
    require_non_overlapping_tensor(*tensor, "RoPE tensor");
  require_floating_data_type(x, "RoPE input must be floating");
  require_same_data_type(x, y, "RoPE output type must match input");
  if (x.dimensions.size() != 4 || y.dimensions != x.dimensions)
    throw ApiError(FLAGDNN_STATUS_INVALID_VALUE,
                   "RoPE requires BHSD input and matching output");
  auto width = integer_attribute(operation, "rope_dim");
  if (width == 0) width = x.dimensions[3];
  if (width < 2 || width % 2 || width > x.dimensions[3])
    throw ApiError(
        FLAGDNN_STATUS_INVALID_VALUE,
        "RoPE rotation width must be even, positive and no larger than D");
  if (freqs.data_type != FLAGDNN_DATA_FLOAT32 ||
      freqs.dimensions !=
          std::vector<std::int64_t>{x.dimensions[2], 1, 1, width})
    throw ApiError(FLAGDNN_STATUS_INVALID_VALUE,
                   "RoPE frequencies must be FP32 [S,1,1,rope_dim]");
  const auto scale = real_attribute(operation, "output_scale");
  if (!std::isfinite(scale))
    throw ApiError(FLAGDNN_STATUS_INVALID_VALUE,
                   "RoPE output scale must be finite");
  return {{{"rope_dim", width}, {"n_elements", x.element_count()}},
          {{"output_scale", scale}},
          {}};
}
}  // namespace flagdnn::native

File: backends/hygon/dispatch/position_embedding.py
# Copyright 2026 FlagOS Contributors
# SPDX-License-Identifier: Apache-2.0

"""Layout validation and launch selection for rotary position embeddings."""

import math
from .extended_common import (
    FLOAT_DATA_TYPES,
    TRITON_POINTER_TYPES,
    _has_non_overlapping_strides,
    _require_integer,
)


def _rope_kernel_configuration(operation, parameters, tensors):
    if len(tensors) != 3:
        raise ValueError("RoPE requires input, frequencies and output")
    x, freqs, y = tensors
    if len(x["dimensions"]) != 4 or y["dimensions"] != x["dimensions"]:
        raise ValueError("RoPE requires BHSD input and matching output")
    if (
        x["data_type"] not in FLOAT_DATA_TYPES
        or y["data_type"] != x["data_type"]
    ):
        raise ValueError("RoPE input/output must use the same floating type")
    for tensor in tensors:
        if not _has_non_overlapping_strides(
            tensor["dimensions"], tensor["strides"]
        ):
            raise ValueError("RoPE requires non-overlapping tensor strides")
    batch, heads, sequence, dimension = x["dimensions"]
    width = _require_integer(
        parameters, "rope_dim", minimum=2, maximum=dimension
    )
    if (
        width % 2
        or freqs["dimensions"] != [sequence, 1, 1, width]
        or freqs["data_type"] != "float32"
    ):
        raise ValueError(
            "RoPE frequencies must be FP32 [S,1,1,rope_dim] "
            "with even rotation width"
        )
    elements = _require_integer(parameters, "n_elements")
    if elements != batch * heads * sequence * dimension:
        raise ValueError("RoPE element count is inconsistent")
    scale = parameters.get("output_scale")
    if (
        isinstance(scale, bool)
        or not isinstance(scale, (int, float))
        or not math.isfinite(scale)
    ):
        raise ValueError("RoPE output scale must be finite")
    constants = {
        "ELEMENTS": elements,
        "HEADS": heads,
        "SEQUENCE": sequence,
        "DIMENSION": dimension,
        "ROPE_DIM": width,
        "SCALE": float(scale),
        "F_S": freqs["strides"][0],
        "F_D": freqs["strides"][3],
        "BACKWARD": operation == "rope_backward",
        "BLOCK_SIZE": 256,
    }
    for role, tensor in [("X", x), ("Y", y)]:
        for axis, stride in zip("BHSD", tensor["strides"]):
            constants[role + "_" + axis] = stride
    grid = (elements + 255) // 256
    if grid > 2**31 - 1:
        raise ValueError("RoPE launch grid exceeds Hygon limits")
    return (
        "rope_kernel",
        {
            "x_ptr": TRITON_POINTER_TYPES[x["data_type"]],
            "freqs_ptr": "*fp32",
            "y_ptr": TRITON_POINTER_TYPES[y["data_type"]],
        },
        constants,
        (grid, 1, 1),
        [("tensor", None)] * 3,
    )
